src/wasm2asm.h

```diff
@@ -944,34 +1054,34 @@ Ref Wasm2AsmBuilder::processFunctionBody(Function* func, IString result) {
     }
 
     Ref visitCallIndirect(CallIndirect* curr) {
-      std::string stable = std::string("FUNCTION_TABLE_") + curr->fullType.c_str();
+      // TODO: the codegen here is a pessimization of what the ideal codegen
+      // looks like. Eventually if necessary this should be tightened up in the
+      // case that the argument expression don't have any side effects.
```

Comment:
I see. Reading it carefully now, it looks like we run `simplify-locals-notee-nostructure` after flattening, which may un-flatten it. However, if we run `simplify-locals-nonesting` then flatness will be preserved. It sounds like that will make the work here a lot simpler, so it's probably worth doing, especially if we're already flat anyhow. It may also end up fixing a bunch of bugs we'd otherwise only run into later.

The potential downside is that in some cases nesting might lead to simpler JS code. But we can probably leave that for JS optimization to do.

Reply:
Ok makes sense, should I change the pass ordering instead of these changes?

Reply:
Well, this is a large issue we need to think about. I think we can defer that thinking to later.